### PR TITLE
fix Issue 14596 - Error: e2ir: cannot cast malloc(42u) of type void* …

### PR DESCRIPTION
**Pull request by Kozzi11**
…to type char[]

### CODE REVIEW

**Comment by yebblies** [comment]
Does a similar issue exist for static arrays?

**Comment by Kozzi11** [comment]
Yep, it exist, I will add this fix too

File: test/fail_compilation/fail14596.d [hunk 1]
```diff
@@ -0,0 +1,10 @@
+/*
+TEST_OUTPUT:
+---
+fail_compilation/fail14596.d(9): Error: cannot cast expression malloc(42LU) of type void* to type char[]
+---
+*/
+void main() {
+    import core.stdc.stdlib : malloc;
+    auto ptr = cast(char[]) malloc(42);
+}
```
Comment (9rnsr) [inline test/fail_compilation/fail14596.d]: As a better place, you can add the new test case to `fail_compilation/fail_casting.d`.

Comment (9rnsr) [inline test/fail_compilation/fail14596.d]: You can also remove the dependency to druntime as follows.

```d
void* p = null;
auto arr = cast(char[])p;
```

And using `malloc` function is not good for a platform independent test case. It takes `size_t` parameter, so the integer value `42` will be uint in 32bit target, and ulong in 64bit target. Finally the difference will appear in the error message.

Let's not use it to fix the current auto-tester failure.

**Comment by WalterBright** [comment]
Auto-merge toggled on

**Pull request merged**